File: src/types/coerce.ts
import { Type, typesEqual, PrimitiveType, OptionalType, ErrorType } from "./type";

export interface CoercionResult {
  ok: boolean;
  type?: Type;
  reason?: string;
}

export function canCoerce(from: Type, to: Type): CoercionResult {
  // Exact match
  if (typesEqual(from, to)) {
    return { ok: true, type: to };
  }

  // Literal to primitive
  if (from.kind === "literal" && to.kind === "primitive") {
    return { ok: true, type: to };
  }

  // T → ?T (wrap in optional)
  if (to.kind === "optional") {
    const inner = canCoerce(from, to.elem);
    if (inner.ok) {
      return { ok: true, type: to };
    }
  }

  // ?T → ?U if T → U
  if (from.kind === "optional" && to.kind === "optional") {
    const inner = canCoerce(from.elem, to.elem);
    if (inner.ok) {
      return { ok: true, type: to };
    }
  }

  // !T → !U if T → U (error propagation)
  if (from.kind === "error" && to.kind === "error") {
    const inner = canCoerce(from.elem, to.elem);
    if (inner.ok) {
      return { ok: true, type: to };
    }
  }

  // No silent crossing between ?T and !T
  if ((from.kind === "optional" && to.kind === "error") ||
      (from.kind === "error" && to.kind === "optional")) {
    return { ok: false, reason: "cannot silently convert between optional and error types" };
  }

  // Numeric widening (e.g., integer 32 → integer 64)
  if (from.kind === "primitive" && to.kind === "primitive") {
    if (from.name === to.name && from.width && to.width) {
      if (from.width < to.width) {
        return { ok: true, type: to };
      }
    }
  }

  return { ok: false, reason: `cannot coerce to target type` };
}

File: test/types/coerce.nim
import ../../src/types/[coerce, type]

proc literal(value: string): Type = Type(kind: "literal", value: value)
proc primitive(name: string; width: int): Type =
  Type(kind: "primitive", name: name, width: width)

doAssert canCoerce(literal("18446744073709551615"),
  primitive("unsigned", 64)).ok
doAssert not canCoerce(literal("18446744073709551616"),
  primitive("unsigned", 64)).ok
doAssert canCoerce(literal("340282366920938463463374607431768211455"),
  primitive("unsigned", 128)).ok
doAssert not canCoerce(literal("340282366920938463463374607431768211456"),
  primitive("unsigned", 128)).ok
doAssert canCoerce(literal("-170141183460469231731687303715884105728"),
  primitive("integer", 128)).ok
doAssert not canCoerce(literal("-170141183460469231731687303715884105729"),
  primitive("integer", 128)).ok
doAssert canCoerce(literal("9007199254740992"),
  primitive("decimal", 64)).ok
doAssert not canCoerce(literal("9007199254740993"),
  primitive("decimal", 64)).ok
doAssert canCoerce(primitive("unsigned", 32),
  primitive("integer", 64)).ok

echo "type coercion parity: ok"
